src/Data/Aeson/Commit.hs

```diff
@@ -23,11 +23,24 @@ commit pre post = Commit $ do
   a <- ExceptT $ captureError pre -- Lift pre's error to the ExceptT level
   lift $ post a
     where
-      captureError :: Parser b -> Parser (Either (Parser Void) b)
-      captureError p = Right <$> p <|> pure (Left $ (const undefined) <$> p)
+      captureError :: Parser b -> Parser (Either [Parser Void] b)
+      captureError p = Right <$> p <|> pure (Left [fmap (const undefined) p])
 
 runCommit :: Commit a -> Parser a
-runCommit (Commit f) = runExceptT f >>= either (fmap absurd) pure
+runCommit (Commit f) = runExceptT f >>= either handleErrors pure
+  where
+  handleErrors :: [Parser Void] -> Parser a
+  handleErrors []     = fail "No parsers tried"
+  handleErrors (p:ps) = fmap absurd (go (p:ps) [] [])
+    where
+    go [] path errors = parserThrowError path ("No match,\n" <> unlines (fmap ("- " <>) errors))
+    go (y:ys) _ msgs = parserCatchError y $ \path msg ->
+      go ys path (msg:msgs)
+        -- TODO: how do we handle the multiple JSONPaths?
+        -- Right now the rightmost failure's path is used when presenting
+        -- the error message. Ideally one path per error would be preferable but
+        -- `aeson` doesn't support such a thing. When errors are reported in `aeson`
+        -- a single JSONPath defines how the error message is presented.
```

Comment:
Not super happy about this, but let's wait until we run into a good error case to use as a reference

Reply:
One thing that I was considering was to collect the longest common prefix among all of the JSONPaths, use that as the JSONPath for the whole error and at each error message output the path relative to the common path. What do you think? Would this become confusing for users?

Reply:
I have a hard time picturing what the current or ideal behavior looks like tbh, could you construct an example? Or just merge it, I think the cases we care about look fine now

Reply:
I'll go ahead and merge this after fixing up the `liftParser` name.

I think we want to look at a more intricate definition of parse errors in a different PR.